src/flashcart/ed64/ed64_internal.h

```diff
@@ -0,0 +1,167 @@
+#define PI_BSD_DOM2_PWD_REG	(PI_BASE_REG+0x28)    /*   pulse width */
+
+/* PI dom2 page size (R/W): [3:0] domain 2 device page size */
+#define PI_BSD_DOM2_PGS_REG	(PI_BASE_REG+0x2C)    /*   page size */
+
+/* PI dom2 release (R/W): [1:0] domain 2 device R/W release duration */
+#define PI_BSD_DOM2_RLS_REG	(PI_BASE_REG+0x30)    /*   release duration */
+
+#define	PHYS_TO_K1(x)	((uint32_t)(x)|0xA0000000)	/* physical to kseg1 */
+#define	IO_WRITE(addr,data)	(*(volatile uint32_t *)PHYS_TO_K1(addr)=(uint32_t)(data))
+#define	IO_READ(addr)		(*(volatile uint32_t *)PHYS_TO_K1(addr))
+
+typedef enum {
+    BI_SPI_SPD_50 = 0,
+    BI_SPI_SPD_25 = 1,
+    BI_SPI_SPD_LO = 2,
+} ed64_spi_speed_t;
+
+#define SPI_CFG_SPD0 1
+#define SPI_CFG_SPD1 2
+#define SPI_CFG_SS 4
+#define SPI_CFG_RD 8
+#define SPI_CFG_DAT 16
+#define SPI_CFG_1BIT 32
+
+/* Initialization functions */
+uint8_t ed64_bios_init();
+void ed64_bios_init_v2();
+void ed64_bios_init_v3();
+
+/* Device information functions */
+uint16_t ed64_bios_get_fpga_version();
+uint16_t ed64_bios_get_cpld_version();
+
+/* Firmware update functions */
+void ed64_bios_load_firmware(uint8_t *firmware);
+
+/* USB functions */
+uint8_t ed64_bios_usb_read_busy();
+uint8_t ed64_bios_usb_read(uint32_t start_address, uint32_t slen);
+uint8_t ed64_bios_usb_write(uint32_t start_address, uint32_t slen);
+
+/* SPI functions */
+uint8_t ed64_bios_spi(uint8_t data);
+void ed64_bios_spi_nr(uint8_t data);
+void ed64_bios_set_spi_speed(uint16_t speed);
+
+/* SD Card functions */
+void ed64_bios_sd_mode(uint16_t mode);
+uint8_t ed64_bios_spi_read_to_rom(uint32_t start_address, uint16_t slen);
+void ed64_bios_byteswap_on();
+void ed64_bios_byteswap_off();
+
+/* Save functions */
+void ed64_bios_set_ram_bank(uint8_t bank);
+ed64_save_type_t ed64_bios_get_save_type();
+void ed64_bios_set_save_type(ed64_save_type_t type);
+
+/* reads metadata related to the assembily date and cart capabilities */
+void ed64_bios_read_cart_metadata(void *dest);
+
+uint16_t ed64_bios_msg_rd();
+void ed64_bios_msg_wr(uint16_t val);
+
+/* DMA functions */
+void ed64_bios_dma_read_rom(void *ram, uint32_t start_address, uint32_t slen);
+void ed64_bios_dma_write_rom(void *ram, uint32_t start_address, uint32_t slen);
+void ed64_bios_dma_read_sram(void *ram, uint32_t address, uint32_t length);
+void ed64_bios_dma_write_sram(void *ram, uint32_t address, uint32_t length);
+
+void ed64_bios_lock_regs();
+void ed64_bios_unlock_regs();
+
+uint32_t ed64_bios_reg_read(uint32_t reg);
+void ed64_bios_reg_write(uint32_t reg, uint32_t data);
+void ed64_bios_reset_spx();
+
+/* GPIO functions */
+void ed64_bios_gpio_mode_rtc();
+void ed64_bios_gpio_mode_io();
+void ed64_bios_gpio_mode_off();
+uint8_t ed64_bios_gpio_read();
+
+/* 64DD cart conversion save functions */
+void ed64_bios_64dd_ram_oe();
+void ed64_bios_64dd_ram_we();
+void ed64_bios_64dd_ram_off();
+void ed64_bios_64dd_ram_clear();
+uint8_t ed64_bios_get_64dd_ram_supported();
```

Comment:
That's interesting, I have no idea what this does and need to be educated what is ED64 for 64DD ROM conversions.

Reply:
My understanding is that the FPGA firmware (for the V3, if updated to a version that supports it) has a list of the writable areas of the DD conversion ROM. When you hot restart after playing the game, it will overwrite the ROM with the changes.

This code has never seen the light of day before, so will probably need me to search through the code base to add the extra bits.

Reply:
Of course, you could just write the whole file back from SDRAM to the SD card, but the use of the dd functions above make this a lot quicker as it only writes the areas needed.

Reply:
Pretty dure that's how 64DD is handled in v2/clones+altra64. It takes a good 7 minutes to save progress. With F-Zero 64DD being an exception if you change the save type and just allow it to save everything other than custom tracks.